src/decode.rs

```diff
@@ -3682,1522 +3693,119 @@ unsafe fn decode_b(
-                    as *mut alias32))
-                    .u32_0 =
-                    (0x1010101 as libc::c_uint)
-                        .wrapping_mul(*(&t_dim.lw as *const uint8_t as *mut uint8_t).offset(1)
-                            as libc::c_uint);
-                (*(&mut *(t.l.mode).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 =
-                    (0x1010101 as libc::c_uint).wrapping_mul(y_mode_nofilt_0 as libc::c_uint);
-                (*(&mut *(t.l.pal_sz).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = (0x1010101 as libc::c_uint)
-                    .wrapping_mul(b.c2rust_unnamed.c2rust_unnamed.pal_sz[0] as libc::c_uint);
-                (*(&mut *(t.l.seg_pred).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = (0x1010101 as libc::c_uint).wrapping_mul(seg_pred as libc::c_uint);
-                (*(&mut *(t.l.skip_mode).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = 0 as libc::c_int as uint32_t;
-                (*(&mut *(t.l.intra).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = 0x1010101 as libc::c_uint;
-                (*(&mut *(t.l.skip).as_mut_ptr().offset(by4 as isize) as *mut uint8_t
-                    as *mut alias32))
-                    .u32_0 = (0x1010101 as libc::c_uint).wrapping_mul(b.skip as libc::c_uint);
-                (*(&mut *(*(t.pal_sz_uv).as_mut_ptr().offset(1))
-                    .as_mut_ptr()
-                    .offset(by4 as isize) as *mut uint8_t as *mut alias32))
-                    .u32_0 = (0x1010101 as libc::c_uint).wrapping_mul(
-                    (if has_chroma {
-                        b.c2rust_unnamed.c2rust_unnamed.pal_sz[1] as libc::c_int
-                    } else {
-                        0 as libc::c_int
-                    }) as libc::c_uint,
+
+        // update contexts
+
+        let y_mode_nofilt = if b.y_mode() == FILTER_PRED as u8 {
+            DC_PRED as IntraPredMode
+        } else {
+            b.y_mode() as IntraPredMode
+        };
+
+        let mut set_ctx = |dir: &mut BlockContext, diridx, off, mul, rep_macro: SetCtxFn| {
+            // NOTE: This corresponds to the following logic in the original C:
+            //
+            // ```
+            // ((uint8_t *) &t_dim->lw)[diridx]
+            // ```
+            //
+            // The original logic casts a pointer to the `lw` field to a `*const u8` and
+            // then does indexing within the struct's fields. This is deeply cursed and
+            // unnecessary since `diridx` will only ever be 0 or 1. So for the Rust version
+            // we manually select which field we want based on `diridx` directly.
+            let lw_lh = match diridx {
```

Comment:
This is indeed quite cursed.  I'm surprised they didn't at least use a `union` with an array, which they did for the same thing in other places.

Maybe just `l` is a better name than `lw_lh`?  Though `lw_hw` is also clear enough I think.